src/vterm.c

```diff
@@ -0,0 +1,474 @@
+  }
+}
+
+size_t get_col_offset(Term *term, int row, int end_col) {
+  int col = 0;
+  size_t offset = 0;
+  unsigned char buf[4];
+
+  while (col < end_col) {
+    VTermScreenCell cell;
+    fetch_cell(term, row, col, &cell);
+    if (cell.chars[0]) {
+      if (cell.width > 1) {
+        offset += cell.width - 1;
+      }
+    }
+    col += cell.width;
+  }
+  return offset;
+}
+
+static bool compare_cells(VTermScreenCell *a, VTermScreenCell *b) {
+  bool equal = true;
+  equal = equal && (a->fg.red == b->fg.red);
+  equal = equal && (a->fg.green == b->fg.green);
+  equal = equal && (a->fg.blue == b->fg.blue);
+  equal = equal && (a->bg.red == b->bg.red);
+  equal = equal && (a->bg.green == b->bg.green);
+  equal = equal && (a->bg.blue == b->bg.blue);
+  equal = equal && (a->attrs.bold == b->attrs.bold);
+  equal = equal && (a->attrs.underline == b->attrs.underline);
+  equal = equal && (a->attrs.italic == b->attrs.italic);
+  equal = equal && (a->attrs.reverse == b->attrs.reverse);
+  equal = equal && (a->attrs.strike == b->attrs.strike);
+  return equal;
+}
+
+Lisp_Object
+allocate_vterm () {
+  Term *term = malloc(sizeof(Term));
+  Lisp_Object val = allocate_misc (Lisp_Misc_Save_Value);
+
+  struct Lisp_Save_Value *p = XSAVE_VALUE (val);
+
+  p->save_type = SAVE_POINTER;
+  p->data[0].pointer = term;
+
+  global_term = term;
+  
+  return val;
+}
+
+Lisp_Object
+refresh_lines (Term *term, int start_row, int end_row, int end_col) {
+
+  int i, j;
+
+  char buffer[((end_row - start_row + 1) * end_col) * 4];
+  int length = 0;
+  VTermScreenCell cell;
+  VTermScreenCell lastCell;
+  fetch_cell(term, start_row, 0, &lastCell);
+
+  int offset = 0;
+  for (i = start_row; i < end_row; i++) {
+    for (j = 0; j < end_col; j++) {
+      fetch_cell(term, i, j, &cell);
+
+      if (!compare_cells(&cell, &lastCell)) {
+        Lisp_Object text = vterm_render_text(buffer, length, &lastCell);
+        call1(intern ("insert"), text);
+        length = 0;
+      }
+      
+      lastCell = cell;
+      if (cell.chars[0] == 0) {
+        buffer[length] = ' ';
+        length++;
+      } else {
+        unsigned char bytes[4];
+        size_t count = codepoint_to_utf8_c(cell.chars[0], bytes);
+        for (int k = 0; k < count; k++) {
+          buffer[length] = bytes[k];
+          length++;
+        }
+      }
+
+      if (cell.width > 1) {
+        int w = cell.width - 1;
+        offset += w;
+        j = j + w;
+      }
+    }
+
+    buffer[length] = '\n';
+    length++;
+  }
+  Lisp_Object text = vterm_render_text(buffer, length, &lastCell);
+  call1(intern ("insert"), text);
+
+  return text; 
+}
+
+VTermScreenCallbacks vterm_screen_callbacks = {
+    .damage = vterm_damage,
+    .moverect = vterm_moverect,
+    .movecursor = vterm_movecursor,
+    /* .settermprop = term_settermprop, */
+    .resize = vterm_resize,
+    .sb_pushline = term_sb_push,
+    .sb_popline = term_sb_pop,
+    
+};
+
+void
+vterm_sb_buffer_size () {
```

Comment:
I'd suggest not making new non-static functions in the `vterm_...` namespace; in case `libvterm` itself starts adding more.

Reply:
True. I wanted to avoid confusion regarding term and vterm, but I will change it to something different.